### PR TITLE
SAK-42944 Recover rubrics with softly deleted assessments

### PR DESCRIPTION
**Pull request by bgarciaentornos**
https://jira.sakaiproject.org/browse/SAK-42944

Some comments:

- As commented on the jira, this will unlock the Rubric on soft deletion (if there aren't more associations). Then all the associations will be removed when removing the Rubric.

- Could the associations be retrieved on deletion time avoiding the ajax call?

- Can't really tell if both ResourceProcessor files are necessary anymore, they come from the original integration but I don't see how they'd be needed nowadays. If someone believes they're still necessary, then I'll have to update them.

### CODE REVIEW

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- rubrics/api/src/main/java/org/sakaiproject/rubrics/logic/model/Rubric.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=4788968&bid=15946492)

**Comment by mpellicer** [comment]
This PR looks good to me, good job @bgarciaentornos.

@ern @adrianfish ?

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java [hunk 1]
```diff
@@ -813,6 +813,87 @@ public void deleteRubricAssociationsByItemIdPrefix(String itemId, String toolId)
...
+			builder.withHeaders(headers);
+
+			Map<String, Object> parameters = new HashMap<>();
+			parameters.put("toolId", toolId);
+			parameters.put("itemId", itemId);
+
+			Resources<Resource<ToolItemRubricAssociation>> associationResources = builder.withTemplateParameters(parameters).toObject(resourceParameterizedTypeReference);
+
+			for (Resource<ToolItemRubricAssociation> associationResource : associationResources) {
+                String associationHref = associationResource.getLink(Link.REL_SELF).getHref();
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java]: Indentation

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java [hunk 2]
```diff
@@ -813,6 +813,87 @@ public void deleteRubricAssociationsByItemIdPrefix(String itemId, String toolId)
...
+                deleteRubricEvaluationsForAssociation(associationHref, toolId);
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java]: If this is a soft delete, should we be deleting the evaluations? If we then recover the associated item, won't we want to get the evaluations back, too?

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java [hunk 3]
```diff
@@ -813,6 +813,87 @@ public void deleteRubricAssociationsByItemIdPrefix(String itemId, String toolId)
...
+			headers.add("Authorization", String.format("Bearer %s", generateJsonWebToken(toolId)));
+			builder.withHeaders(headers);
+
+			Map<String, Object> parameters = new HashMap<>();
+			parameters.put("toolId", toolId);
+			parameters.put("itemId", itemId);
+
+			Resources<Resource<ToolItemRubricAssociation>> associationResources = builder.withTemplateParameters(parameters).toObject(resourceParameterizedTypeReference);
+			for (Resource<ToolItemRubricAssociation> associationResource : associationResources) {
+                String associationHref = associationResource.getLink(Link.REL_SELF).getHref();
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/logic/RubricsServiceImpl.java]: Some more indentation

File: rubrics/tool/src/main/frontend/webcomponents/sakai-item-delete.js [hunk 4]
```diff
@@ -137,6 +137,27 @@ export class SakaiItemDelete extends RubricsElement {
...
+      if (data._embedded['rubric-associations'].length) {
+        data._embedded['rubric-associations'].forEach( assoc => {					
+          var assocUrl = "/rubrics-service/rest/rubric-associations/"+assoc.id;
+          $.ajax({
+		    url: assocUrl,
+		    method: "DELETE",
+		    headers: {"authorization": this.token},
+		    contentType: "application/json"
+		  }).fail((jqXHR, error, message) => {
+		    console.log(error);
```
Comment (adrianfish) [inline rubrics/tool/src/main/frontend/webcomponents/sakai-item-delete.js]: Perhaps use console.error here?

**Comment by bgarciaentornos** [comment]
> Generally, the approach looks good. I have a concern about the evaluations being deleted though. What happens if you delete an assessment that has been graded, now? Do you get a warning or something? I'd also prefer that the deletion of the soft deleted associations happened on the backend during the rubric deletion op, but ... I'm in the middle of removing traverson and spring boot from the backend so that can get refactored later.

I believe you can't remove an assessment if you don't remove all the submissions first.

**Comment by steveswinsburg** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 11
           

Complexity increasing per file
==============================
- rubrics/api/src/main/java/org/sakaiproject/rubrics/logic/model/Rubric.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/sakaiproject/sakai/pullRequest?prid=4788968&bid=15946492)

**Comment by ern** [comment]
Are assessment submissions deleted? That doesn't seem to be inline with the concept of soft delete (where nothing is deleted it's just marked for deletion this way when you undelete everything is intact)

**Comment by ern** [comment]
Currently assessment submissions are deleted on a soft delete and this PR brings rubrics evaluations consistent with this.

**Pull request merged**